src/frontend/src/binder/expr/mod.rs

```diff
@@ -363,3 +358,29 @@ pub fn bind_data_type(data_type: &AstDataType) -> Result<DataType> {
     };
     Ok(data_type)
 }
+
+/// Find the `least_restrictive` type over a list of `exprs`, and add implicit cast when necessary.
+/// Used by `VALUES`, `CASE`, `UNION`, etc. See [PG](https://www.postgresql.org/docs/current/typeconv-union-case.html).
+pub fn align_types<'a>(exprs: impl Iterator<Item = &'a mut ExprImpl>) -> Result<DataType> {
+    use std::mem::swap;
+
+    let exprs = exprs.collect_vec();
+    // Essentially a filter_map followed by a try_reduce, which is unstable.
+    let mut ret_type = None;
+    for e in &exprs {
+        if e.is_null() {
+            continue;
+        }
+        ret_type = match ret_type {
+            None => Some(e.return_type()),
+            Some(t) => Some(least_restrictive(t, e.return_type())?),
+        };
+    }
+    let ret_type = ret_type.unwrap_or(DataType::Varchar);
```

Comment:
```
    let ret_type = exprs.iter().filter_map(|e| {
        match e.is_null() {
            true => None,
            false => Some(e.return_type()),
        }
    }).try_reduce(least_restrictive)?.unwrap_or(DataType::Varchar);
```
@skyzh What criteria do we use to decide if an unstable feature is okay to enable?
```
error[E0658]: use of unstable library feature 'iterator_try_reduce': new API
   --> src/frontend/src/binder/expr/mod.rs:373:8
    |
373 |     }).try_reduce(least_restrictive)?.unwrap_or(DataType::Varchar);
    |        ^^^^^^^^^^
    |
    = note: see issue #87053 <https://github.com/rust-lang/rust/issues/87053> for more information
    = help: add `#![feature(iterator_try_reduce)]` to the crate attributes to enable
```

Reply:
I believe this is covered by itertools?

Reply:
Unfortunately no 😢

Reply:
I think this can be easily implemented with `try_fold`? For the `try_reduce` feature, I didn't see it's actively developed.

Reply:
Yes. But it still requires folding `None` manually, which is the main reason this version is shorter than the current one in PR ...

Reply:
e.g. try_fold(None, <the original function>)

Reply:
Well, then I have no strong opinion on this... I'd be okay with any approach 🤣

Reply:
I will keep the current one then 😂